Add metrics for caching

maryamariyan opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/50406

Adds the approved APIs:

```c#
    public partial class MemoryCache
    {
        /// <summary>
        /// Gets a snapshot of the current statistics for the memory cache.
        /// </summary>
        public MemoryCacheStatistics? GetCurrentStatistics() { throw null; }
    }

    public partial interface IMemoryCache
    {
#if NET_7_OR_GREATER
        /// <summary>
        /// Gets a snapshot of the cache statistics if available.
        /// </summary>
        /// <returns>An instance of the <see cref="MemoryCacheStatistics"/> instance.</returns>
        public MemoryCacheStatistics? GetCurrentStatistics() => null;
#endif
    }

    public partial class MemoryCacheStatistics
    {
        public MemoryCacheStatistics() {}
        public long TotalMisses { get; init; } 
        public long TotalHits { get; init; }
        public long CurrentEntryCount { get; init; }
        public long? CurrentEstimatedSize { get; init; }
    }
```

Also:
```diff
    public partial class MemoryCacheOptions : Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Caching.Memory.MemoryCacheOptions>
    {
+        public bool TrackStatistics { get { throw null; } set { } }
    }
```

In this implementation, the Size is not in sync with the backing count from CoherentState's concurrent dictionary count, which means that the snapshot of the statistics may be out of sync when taken during heavy multithreaded load is happening, but the values will eventually become consistent once the heavy load is finished. (refer to the last test case added)

TODO:
- The PR is ready for review, in the meantime I am going to make sure we are not causing regression for tech empower tests, and will share the results here

---

### Alternative Approaches

- If we wanted to be completely accurate with the snapshot values, returning count/size that are in sync, then I also had prototyped this change https://github.com/dotnet/runtime/pull/66479/commits/a68d3e4b9494fbaec1be7d2609bf55ba725fe84d by adding a separate `_cacheCount` that would always remain in sync with `_cacheSize`. To keep things simple I decided to go with the easiest fix drafted here in the PR instead which just returns the estimated size.

- Another approach that could be explored would be to use a thread static object (e.g. `ThreadStats` that keeps track of count/size). The actual count/size would be the aggregated value of `ThreadStats` count/size for all of the threads present. Disadvantage with using this thread static approach is that since that way we are keeping track of a list of threads, we need additional code to keep track of threads that have gone stale over time. We could explore this option too later if we learn later that the current solution in the PR is not sufficient or performant enough.

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.cs:
```diff
@@ -126,6 +126,21 @@ public partial class MemoryCacheEntryOptions
     }
+    public partial struct MemoryCacheStatistics
+    {
+// #if NETCOREAPP
```
will you keep these comments?

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.cs:
```diff
@@ -126,6 +126,21 @@ public partial class MemoryCacheEntryOptions
+//         public long TotalRequests { get { throw null; } init { throw null; } }
+// #else
+        private int _dummyPrimitive;
+        public long CurrentEntryCount { readonly get { throw null; } set { } }
```
>set [](http://example.com/codeflow?start=69&length=3)

I recall in the review decided to not expose the setters. right?

tarekgh replied:
If you are going to enable init, please look at the PR [https://github.com/dotnet/runtime/pull/65530/files](https://github.com/dotnet/runtime/pull/65530/files) to see how we handled it and ensure you'll have the typeforward there.

cc @ericstj

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -196,6 +199,24 @@ internal void SetEntry(CacheEntry entry)
+            {
+                Interlocked.Increment(ref _totalRequests);
+                return false;
+            }
```
May be can be written like:

```C#
            Interlocked.Increment(ref _totalRequests);
            if (TryGetValueInternal(key, out result))
            {
                    Interlocked.Increment(ref _totalHits);
                    return true;
            }
             return false;
```

MihaZupan commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -27,6 +27,9 @@ public class MemoryCache : IMemoryCache
         private CoherentState _coherentState;
         private bool _disposed;
+        private readonly object _sync = new object();
+        private long _totalHits;
+        private long _totalRequests;
```
A trick you can use it to store the number of hits & misses instead, saving one `Interlocked` operation on the hot path.
```c#
private long _totalHits;
private long _totalMisses;
```

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.Typeforwards.netcoreapp.cs:
```diff
@@ -0,0 +1,6 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// The compiler emits a reference to the internal copy of this type in our non-NETCoreApp assembly
+// so we must include a forward to be compatible with libraries compiled against non-NETCoreApp System.Diagnostics.DiagnosticSource
```
>System.Diagnostics.DiagnosticSource [](http://example.com/codeflow?start=96&length=35)

you need to change the name of this assembly

maryamariyan committed:
Adds API to retrieve caching metrics

    - Adds MemoryCacheStatistics struct and GetCurrentStatistics API
    - Ignore MatchingRefApiCompat failure on struct properties
    - Implement MemoryCacheStatistics props with init

maryamariyan committed:
Make CurrentSize and CurrentEntryCount return coherent values

maryamariyan committed:
Add test

maryamariyan committed:
Apply API review feedback - TODO: DIM support

maryamariyan committed:
Remove MatchingRefApiCompatBaseline, leftover from init on struct

maryamariyan committed:
Undo CoherentState._cacheSize changes and add DIM

halter73 commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.csproj:
```diff
@@ -6,6 +6,11 @@
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.Caching.Abstractions.cs" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.other.cs" Condition="'$(TargetFramework)' != 'net7.0'" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.net7.cs" Condition="'$(TargetFramework)' == 'net7.0'" />
```
Does this play nicely with `t:GenerateReferenceAssemblySource`? Can we do `#if NET7_0_OR_GREATER` in the reference assembly source so we don't have to change this when we start on `net8.0`?

maryamariyan replied:
I don't see any precedence with using ifdef with ref files. Perhaps the condition here could be improved to also cover not just net7.0 but also higher. Also, I assume it won't be possible to use `\t:GenerateReferenceAssemblySource` alongside this.

/cc @joperezr

tarekgh replied:
I think Stephen Halter idea will work too.

maryamariyan replied:
Just to confirm, I tried using the `GenerateReferenceAssemblySource` and noticed it would add `GetCurrentStatistics` to IMemoryCache on `ref\Microsoft.Extensions.Caching.Abstractions.cs`. That's not exactly what we would want, since we prefer to move this addition to a separate `*.net7.cs` file and include it conditionally:

```diff
    public partial interface IMemoryCache : System.IDisposable
    {
        Microsoft.Extensions.Caching.Memory.ICacheEntry CreateEntry(object key);
+        Microsoft.Extensions.Caching.Memory.MemoryCacheStatistics? GetCurrentStatistics() { throw null; }
        void Remove(object key);
        bool TryGetValue(object key, out object? value);
    }
```

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.csproj:
```diff
@@ -6,6 +6,11 @@
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.Caching.Abstractions.cs" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.other.cs" Condition="'$(TargetFramework)' != 'net7.0'" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.net7.cs" Condition="'$(TargetFramework)' == 'net7.0'" />
```
>'net7.0 [](http://example.com/codeflow?start=108&length=7)

I think you need to use $(NetCoreAppCurrent) instead?

halter73 commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/MemoryCacheGetCurrentStatisticsTests.cs:
```diff
@@ -0,0 +1,182 @@
+            }
+            else
+            {
+                Assert.Null(stats.CurrentEstimatedSize );
```
Is it worth adding a test verifying the `GetCurrentStatistics()` DIM returns null?

maryamariyan replied:
TODO: 

- [ ] Add a test verifying the GetCurrentStatistics() DIM returns null

tarekgh reviewed (approved):
Modulo @halter73 comment, LGTM.

maryamariyan committed:
Add DIM test and fix up csproj condition

tarekgh commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.csproj:
```diff
@@ -6,6 +6,10 @@
 
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.Caching.Abstractions.cs" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.net7.cs" Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))" />
```
CC @ericstj @ViktorHofer

ericstj replied:
We use this pattern a lot when we are making a change that needs to be true for a given framework and all future compatible frameworks.  I think this is a good usage 👍   For other usages you can have a look at https://github.com/dotnet/runtime/search?q=IsTargetFrameworkCompatible

maryamariyan committed:
CommonPath -> CoreLibSharedDir

maryamariyan committed:
ref fix up

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -270,6 +274,17 @@ public void Clear()
             }
         }
 
+        public MemoryCacheStatistics GetCurrentStatistics()
```
public API should have xml doc comments.

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheStatistics.cs:
```diff
@@ -0,0 +1,39 @@
+        public long CurrentEntryCount { get; init; }
+
+        /// <summary>
+        /// A snapshot of size at the current state
```
Do we want to document when this returns `null`?

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheStatistics.cs:
```diff
@@ -0,0 +1,39 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.ComponentModel;
```
Is this using needed?

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheStatistics.cs:
```diff
@@ -0,0 +1,39 @@
+        public MemoryCacheStatistics() { }
+
+        /// <summary>
+        /// A snapshot of entry count at the current state
```
```suggestion
        /// Gets the number of <see cref="ICacheEntry" /> instances currently in the memory cache.
```

maryamariyan committed:
Applies PR feedback

maryamariyan commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/MemoryCacheGetCurrentStatisticsTests.cs:
```diff
@@ -0,0 +1,198 @@
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void GetCurrentStatistics_EntrySizesAreOne_MultithreadedCacheUpdates_CountAndSizeRemainInSync(bool sizeLimitIsSet)
```
Removed the test. I had added this for the previous logic. I don't think it makes sense now

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/IMemoryCache.cs:
```diff
@@ -30,5 +30,13 @@ public interface IMemoryCache : IDisposable
+        /// <summary>
+        /// Gets a snapshot of the cache statistics if available.
+        /// </summary>
+        /// <returns>An instance of the <see cref="MemoryCacheStatistics"/> instance.</returns>
```
```suggestion
        /// <returns>An instance of <see cref="MemoryCacheStatistics"/> containing a snapshot of the cache statistics.</returns>
```

eerhardt reviewed (approved):
LGTM. Nice work here @maryamariyan.

maryamariyan committed:
Update src/libraries/Microsoft.Extensions.Caching.Abstractions/src/IMemoryCache.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

KalleOlaviNiemitalo commented:
In the description, you write

> Adds the approved APIs:
>
> ```C#
>     public static partial class MemoryCache
> ```

but the class is not static.

maryamariyan commented:
> but the class is not static.

updated description it was a typo

maryamariyan commented:
Using crank to evaluate these changes over TE benchmarks I came up with the following results:

<img width="652" alt="image" src="https://user-images.githubusercontent.com/5897654/158927185-8c1808ff-6c08-443d-8bb3-6fd00b117de1.png">

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/5897654/158927223-2051df3f-fae7-4724-9c61-ebe804bb6e50.png">

<img width="687" alt="image" src="https://user-images.githubusercontent.com/5897654/158927258-7c5e21f9-c63b-4c5b-9f72-5617ead85eff.png">

for future reference these were the scripts:

```
> crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml \
    --scenario caching --profile aspnet-perf-win --json mar17-win-approved.json \
    --application.options.outputFiles C:\build\mar17\approved\Microsoft.Extensions.Caching.Abstractions.dll \
    --application.options.outputFiles C:\build\mar17\approved\Microsoft.Extensions.Caching.Memory.dll \
    --application.framework net7.0 --application.options.collectCounters true  \
    --application.sdkversion 7.0.100-preview.3.22167.1 --application.collect true \
    --variable warmup=0 --variable duration=5 --chart

> crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml \
    --scenario caching --profile aspnet-perf-win --json mar17-win-main.json \
    --application.options.outputFiles C:\build\mar17\main\Microsoft.Extensions.Caching.Abstractions.dll \
    --application.options.outputFiles C:\build\mar17\main\Microsoft.Extensions.Caching.Memory.dll \
    --application.framework net7.0 --application.options.collectCounters true \
    --application.sdkversion 7.0.100-preview.3.22167.1 --application.collect true \ 
    --variable warmup=0 --variable duration=5 --chart

> crank compare mar17-win-main.json mar17-win-approved.json
```

where the dlls from `Microsoft.Extensions.Caching.Memory.dll` and `Microsoft.Extensions.Caching.Abstractions.dll` were built:
- From commit e7be224241dd4b095d697d740757fb25330f5504 the latest commit on this PR for `C:\build\mar17\approved` 
- From commit 198942c9b99e17ef8bf89898ce3bc6db0bad5782 the base of this PR for `C:\build\mar17\main`.

### Overview

Wondering if the requests per seconds drop is acceptable for us to merge this PR as-is. Alternative is to make it opt-in behaviour.

I have also collected the `etl.zip` files will inspect them tomorrow.

eerhardt commented:
@adamsitnik @sebastienros - thoughts on the above Tech Empower results? I'm a bit surprised incrementing a `long` would reduce the throughput by 5%.

ViktorHofer commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.net7.cs:
```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
I would rename this file to `Microsoft.Extensions.Caching.Abstractions.net7.0.cs` as we usually use one the following suffixes whenever we need to condition on target frameworks:
- TargetFrameworkIdentifier: either `netcoreapp`, `netstandard` or `netframework`
- TargetFramework: i.e. `netstandard2.0`, `net462`, `net6.0`, `net7.0`, ...

KalleOlaviNiemitalo commented:
If the extra time is spent in Interlocked.Increment, can this be sped up with per-cpu counters so that each cpu updates a separate cache line most of the time?

maryamariyan commented:
Looking again this morning. Also, checking benchmarks for with approved changes minus the two `Interlocked.Increment` statements in `TryGetValue`:

<img width="769" alt="image" src="https://user-images.githubusercontent.com/5897654/159039161-7fa3c523-99cb-4b23-b489-2e9d338a25f2.png">


From just uncommenting the two interlocked statements we get:

<img width="680" alt="image" src="https://user-images.githubusercontent.com/5897654/159039016-ce5ea7c7-d1ff-4d79-996d-cf772d3cdda6.png">

all other portions of approved PR benchmarked against base commit yield the following:

<img width="651" alt="image" src="https://user-images.githubusercontent.com/5897654/159039446-ed66a40d-5969-4c66-95a5-48037e7947b2.png">

maryamariyan committed:
Rename to "*net7.0.cs" from "*net7.cs"

sebastienros commented:
`--application.collect true --variable warmup=0 --variable duration=5` should only be used to get the native trace. To get numbers for comparison don't run the collection, and don't change the default warmup/duration. 

We should discard the comment that displays the results with this command.

maryamariyan commented:
Thanks @sebastienros. When not collecting the two interlocked statements reduce the RPS by 50% which is what I noticed yesterday as well:

<img width="771" alt="image" src="https://user-images.githubusercontent.com/5897654/159059346-7f4008f8-cbf3-4189-9098-372999afe63f.png">

The other portions of the PR are not a culprit for this. Benchmarking the approved PR without the interlocked statements I get:
<img width="750" alt="image" src="https://user-images.githubusercontent.com/5897654/159059653-de91fc83-b83d-4569-9f0e-ac0c12cd98f5.png">

maryamariyan committed:
using thread local to get per memory cache stats and better perf

maryamariyan commented:
The new commit I just pushed up [77e390d](https://github.com/dotnet/runtime/pull/66479/commits/77e390d792ec1cdef4d0ba5642b775f07326f7ec), changes the implementation for hit/miss computation to use ThreadLocal. This way we could get per memory cache stats.

Here's the perf numbers for TE with/without this new diff:

![image](https://user-images.githubusercontent.com/5897654/159094641-0fe85052-98c4-4d59-98b3-9bf32f4dc77c.png)

![image](https://user-images.githubusercontent.com/5897654/159094614-9ec73fa6-0d73-499c-9932-87f23df20c63.png)

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheStatistics.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <summary>
+        /// Gets an estimated sum of all the <see cref="ICacheEntry.Size" /> values currently in the memory cache.
+        /// </summary>
+        /// <returns>Returns <see langword="null"/> if no size limit is set for the cache.</returns>
```
Nit: We might want to make this comment more flexible to account for other implementations which could have their own policy about when the value is tracked
```suggestion
        /// <returns>Returns <see langword="null"/> if size isn't being tracked. The common MemoryCache implementation tracks size whenever a SizeLimit is set on the cache.</returns>
```

maryamariyan commented:
As a summary to this PR, if the 4% drop in requests per seconds for the caching-specific tech empower benchmarks is acceptable, then given the new changes get approved then we could submit this.

Otherwise, we could look into adding an opt-in flag for MemoryCacheOptions that would guard the `Hit()`/`Miss()` logic.

I think I noticed there was a 1% error threshold for RPS, comparing same benchmarks between today/yesterday. @sebastienros would you say the TE benchmarks are close to real world samples or higher duty on caching than normal?

/cc @eerhardt @tarekgh would you please review again the new diff?

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +313,88 @@ void ScheduleTask(DateTimeOffset utcNow)
+
+        private Stats GetStats()
+        {
+            if (_stats == null)
```
This initialization could be done eagerly in the constructor and it would probably give a slight performance improvement. You would also need to move the code that appends to the _allStats list inside the init function passed to the ThreadLocal.ctor so that it only runs when the per-thread state is created.

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +313,88 @@ void ScheduleTask(DateTimeOffset utcNow)
+            {
+                _memoryCache = memoryCache;
+            }
+            ~Stats()
```
Not something I would expect to be part of this PR (and maybe not ever) but I noticed the ThreadLocal<T> API has support for a Values property which exposes all the current values of the ThreadLocal across all threads. Unfortunately I don't think you can make use of it because there is no callback when threads exit and their values are removed from that list. If the ThreadLocal API ever supported such a callback in the future then the code here could be simplified to make use of it rather than re-creating the list tracking and cleanup mechanics a 2nd time. I certainly haven't thought through all the implications of adding such a callback so maybe it was already considered in the past and rejected with good reason. @kouvel @stephentoub

kouvel replied:
This looks very similar to what [ThreadInt64PersistentCounter](https://github.com/dotnet/runtime/blob/dd3b61ddc1b1a553efbed8bbdba2938a3dff4140/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs) is trying to do, to act as a counter for completed thread pool work items. I had tried using `ThreadLocal<T>` initially but it was too slow and on some typical processors even worse than `Interlocked.Increment`. So I had ended up with that. Accessing a thread-local static variable still appears to be much slower in .NET Core than what I would expect, but at least I got similar perf compared to `Interlocked` and it would probably scale a tiny bit better. If `ThreadLocal<T>` is not being used to track values I wonder if perf would be better to just use a thread-local static variable instead.

danmoseley replied:
> I wonder if perf would be better to just use a thread-local static variable instead, though I guess that may not be feasible here.

You mean [ThreadStatic] is faster than `ThreadLocal<T>`? Why would it not be feasible? Seems like it would work too.

elinor-fung replied:
The statistics are being tracked per `MemoryCache`, so they are tracked on the instance rather than as a static variable.

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +313,88 @@ void ScheduleTask(DateTimeOffset utcNow)
+                        }
+                    }
+                    allStats.RemoveAll(s => s is null);
+                    Interlocked.Add(ref _memoryCache._accumulatedStats.Hits, Hits);
```
Nit: you don't need the interlocked operations because this code is already synchronized by the lock.

noahfalk reviewed (approved):
LGTM, a few suggestions inline

maryamariyan committed:
Applies PR feedback - with micro perf improvement

maryamariyan commented:
With the micro improvements suggested by @noahfalk the request per second drop reduced to 3.29% from 4.92% on the latest commit

<img width="816" alt="image" src="https://user-images.githubusercontent.com/5897654/159100185-60b6b28b-efbb-4782-9dc9-f68f740d6cc7.png">

maryamariyan commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +314,73 @@ void ScheduleTask(DateTimeOffset utcNow)
+
+        internal sealed class Stats
+        {
+            public int Hits;
+            public int Misses;
```
Refer to earlier comment in https://github.com/dotnet/runtime/pull/66479#issuecomment-1072874447

> one limitation to note is the hit/miss values are int rather than long.

stephentoub replied:
If you change them to be long (to avoid overflowing) you will also need to use Interlocked.Write/Read, at least on 32-bit, assuming we care about avoiding torn values. (This will have a perf impact on 32-bit, but will also avoid overflowing.)

noahfalk replied:
Is there a perf concern if they were long? With all the hot path writes being thread-local data I wouldn't expect it to make much difference performance wise and a long running service with heavy cache usage might overflow the int.

stephentoub replied:
Functional correctness is more important then perf, so if long is needed to be functionally correct, so be it. But we might find it adds enough overhead we need to make it opt-in, if we care about 32-bit perf and care about avoiding the possibility of torn reads/writes with whatever thread is monitoring the data. Making them long would require Interlocked.Writes on 32-bit if we want to avoid tears, and that would amount to effectively the overhead of taking an uncontended lock on every get.

noahfalk replied:
oh I was being dense, I see what you mean about the torn read now. Can we give a shot and see what it does to perf? I think we could make at least a weak argument that 32 bit wasn't incorrect, it was simply a known design limitation. The (hopefully small) group of customers who saw wildly wrong hit ratios probably wouldn't be happy with us though.

maryamariyan committed:
nit fix/readability improvement

maryamariyan commented:
One last thing before pausing the work here, I noticed that the "request per seconds" fluctuates between %3.19 and %4.53 on multiple runs.

Will keep the diff as-is until there's more feedback/opinions on the fix.

Thanks

MihaZupan commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +314,82 @@ void ScheduleTask(DateTimeOffset utcNow)
+                {
+                    if (_allStats[i] is WeakReference<Stats> wr && wr.TryGetTarget(out Stats? stats) && stats == current)
+                    {
+                        _allStats[i] = null;
+                        break;
+                    }
+                }
+                _allStats.RemoveAll(s => s is null);
```
```suggestion
                        _allStats.RemoveAt(i);
                        break;
                    }
                }
```

maryamariyan committed:
apply nit feedback

maryamariyan commented:
RE: https://github.com/dotnet/runtime/pull/66479#discussion_r830435416

@stephentoub would it be OK to merge this with the int32 limitation and 2% RPS reduction and track the perf as a follow up work by including an options flag to enable this for users who need this behaviour?

maryamariyan commented:
Based on offline conversation @eerhardt , @sebastienros and @adamsitnik would be ok with 2% drop.

@sebastienros prefers long and having feature opt-in. I could address this as a follow up.

stephentoub commented:
> Based on offline conversation @eerhardt , @sebastienros and @adamsitnik would be ok with 2% drop.

I think the bigger issue is the correctness one.  Are we ok that this can (easily?) overflow?  Seems numbers would start to be really wonky and incomparable across hits and misses.

I suggest we start from a position of "these need to be 64-bit if we care about correctness", do what's necessary for correctness, and then make it opt-in if we can't get the perf where it needs to be by default.

danmoseley commented:
> @sebastienros prefers long and having feature opt-in. I could address this as a follow up.

My 2c, I would rather not commit a change that drops TE with a plan to get it back. It's too important to add "debt" as we work. Either we're OK with a TE drop, and we can commit, or we should make a single change that does not drop it.

maryamariyan committed:
supporting long hit/miss

maryamariyan committed:
Adds MemoryCacheOptions.TrackStatistics options flag

maryamariyan commented:
I marked this PR as `NO-MERGE` until it receives API approval for the following API changes to support opt-in capability.

```diff
    public partial class MemoryCache : Microsoft.Extensions.Caching.Memory.IMemoryCache, System.IDisposable
    {
-        public Microsoft.Extensions.Caching.Memory.MemoryCacheStatistics GetCurrentStatistics() { throw null; }
+        public Microsoft.Extensions.Caching.Memory.MemoryCacheStatistics? GetCurrentStatistics() { throw null; }
    }
    public partial class MemoryCacheOptions : Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Caching.Memory.MemoryCacheOptions>
    {
+        public bool TrackStatistics { get { throw null; } set { } }
    }
```

Given the API gets approved the PR should then be ready for review.

halter73 commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -270,6 +279,26 @@ public void Clear()
 
+        /// <summary>
+        /// Gets a snapshot of the current statistics for the memory cache.
+        /// </summary>
```
```suggestion
        /// </summary>
        /// <returns>Returns <see langword="null"/> if statistics are not being tracked because <see cref="MemoryCacheOptions.TrackStatistics" /> is <see langword="false"/>.</returns>
```

maryamariyan committed:
Update src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

noahfalk commented:
> I marked this PR as NO-MERGE until it receives API approval for the following API changes to support opt-in capability

I thought we were going to first measure perf to see if we needed such an API?

danmoseley commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,93 @@ void ScheduleTask(DateTimeOffset utcNow)
+                {
+                    if (_allStats![i] is WeakReference<Stats> wr && wr.TryGetTarget(out Stats? stats) && stats == current)
+                    {
+                        _allStats!.RemoveAt(i);
```
I'm assuming that _allStats doesn't get very big? Since the backing storage of the list will never shrink

maryamariyan replied:
I don't think it would grow very big but I'll investigate to see if we need to set capacity for it

maryamariyan replied:
_allStats size should grow at most to the number of cores available.

noahfalk replied:
ThreadLocal is per-thread rather than per-core, but still agree the size shouldn't be an issue.

stephentoub replied:
It should effectively be the max concurrent number of threads that have been alive at the same time (assuming they all access the cache). I'm not concerned about list size, but if we want we could TrimExcess after summing.

maryamariyan replied:
Added TrimExcess

maryamariyan commented:
> I thought we were going to first measure perf to see if we needed such an API?

I checked how the perf gets impacted by using long+threadlocal+interlocked that caused a 5% drop which I assumed would not be within acceptable range.

maryamariyan committed:
update on lock for 32bit machine

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/ref/Microsoft.Extensions.Caching.Memory.cs:
```diff
@@ -45,6 +46,7 @@ public partial class MemoryCacheOptions : Microsoft.Extensions.Options.IOptions<
         Microsoft.Extensions.Caching.Memory.MemoryCacheOptions Microsoft.Extensions.Options.IOptions<Microsoft.Extensions.Caching.Memory.MemoryCacheOptions>.Value { get { throw null; } }
         public long? SizeLimit { get { throw null; } set { } }
         public bool TrackLinkedCacheEntries { get { throw null; } set { } }
+        public bool TrackStatistics { get { throw null; } set { } }
```
what do people think about making the default value of this property true? I suspect most developers, even those writing high performance services, will still probably prefer being able to monitor their cache statistics rather than squeezing out a tiny bit more speed.

maryamariyan replied:
during the API review meeting today the ask was to disable it by default to not hurt the perf for those not tracking the stats

maryamariyan replied:
Just tested it again now, with TrackStatistics enabled. 

As summarized earlier, the new approach would drop the TE RPS by 5%:

![image](https://user-images.githubusercontent.com/5897654/159603106-fc2120a5-d02b-466b-8d2b-b13b66ae6336.png)

noahfalk replied:
IMO true would be the better default but I'm not going to die on that hill : )

davidfowl replied:
We can disable stats in the benchmark though, this is similar to what we do with logging.  That should apply here as well right?

halter73 replied:
I think the difference is that far more people want logging by default than will want `MemoryCache.GetCurrentStatistics()` by default. On the other hand as @eerhardt said in API review, this benchmark is pretty much the worst-case overhead.

I think defaulting to not tracking makes sense, but I'm not super opinionated.

maryamariyan replied:
I'd be fine either way, though this would need another API review churn. Would prefer to have this PR merged as-is and park this discussion on a separate issue if OK.

noahfalk replied:
This would be my speculation how projects break down:

1. Projects that deploy on x86 and run within one order of magnitude of the load created by the TE benchmark: < 0.1% of apps
2. Projects where the developer wants to directly invoke GetCurrentStatistics(): <1% of apps
3. Projects where the developer will appreciate using a monitoring solution (AppInsights, OpenTelemetry, etc) or ad-hoc tools (dotnet-counters) to make the stats available for them in the future: 10-20% apps
4. Projects where neither the stats nor the perf matters: 50+% of apps

I'm worried about the experience of people in group (3). I want them to be able to turn on telemetry very easily when they want it. Anything where they have to follow extra instructions or change more lines of code creates friction in that process. If that means a tiny set of projects in group (1) have to have some friction to disable stats they don't care about I think it is the right tradeoff given my guess at the relative sizes of those groups. Certainly if others think my guesses about the group sizes are wrong they might come to a different conclusion.

maryamariyan replied:
I'd assume for case 3, when using dotnet-counters there would still be need for code change for setting up counters with the GetCurrentStatistics API and in which case also setting the flag should be not too much churn.

But thinking perhaps with OT or AppInsights there is no need for code change and therefore, as you said once those solutions are in for cache metrics retrieval via OT, then it could be that extra instruction for them to also set TrackStatistics to true.

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,113 @@ void ScheduleTask(DateTimeOffset utcNow)
+            long hits = 0, misses = 0;
+            lock (_allStats!)
+            {
+                if (IntPtr.Size == 4)
```
```suggestion
        hits += _accumulatedStats!.Hits;
        misses += _accumulatedStats!.Misses;
```
The local variables are never accessed from concurrent threads and the _accumulatedStats fields are only accessed under the _allStats lock so no need for interlocked here.

noahfalk commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,113 @@ void ScheduleTask(DateTimeOffset utcNow)
+
+                if (IntPtr.Size == 4)
+                {
+                    Interlocked.Add(ref _accumulatedStats!.Hits, Interlocked.Read(ref current!.Hits));
```
```suggestion
                    _accumulatedStats!.Hits += Interlocked.Read(ref current.Hits);
                    _accumulatedStats!.Misses += Interlocked.Read(ref current.Misses);
```
accumulatedStats are never modified outside the lock so no interlocked is necessary there.

maryamariyan replied:
would I still not need interlocked because of preventing torn state on 32bit?

halter73 replied:
Do we even need the `Interlocked.Read` in this case where it's only called from `current`'s finalizer? That means there can't be any thread concurrently modifying this, right?

I don't think we need to worry about any 32 bit tearing if there's no concurrent access and all writes have been committed which I would expect to be the case in a finalizer.

maryamariyan replied:
The only other place the Misses and Hits are changed for Stats is in `Hit()` and `Miss()`

KalleOlaviNiemitalo replied:
OTOH, if it's a finalizer of a thread-local object, then I expect it will be called so rarely that hardly any CPU time will be spent there.

danmoseley replied:
Why is Interlocked.Read necessary on a long when in a 64 bit process? Can't we make it conditional on IntPtr.Size == 4?

noahfalk replied:
Although it isn't necessary it also shouldn't be problematic because this code isn't on the hot path. I doubt we'd see any perf benefit in exchange for adding the conditional.

maryamariyan committed:
undo unnecessary Interlock

maryamariyan committed:
Add TrimExcess

maryamariyan committed:
test cleanup

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -53,6 +53,13 @@ public MemoryCache(IOptions<MemoryCacheOptions> optionsAccessor!!, ILoggerFactor
+            {
+                _allStats = new();
+                _accumulatedStats = new Stats(this, skipAdd: true);
+                _stats = new ThreadLocal<Stats>(() => new Stats(this));
```
What happens if someone sets TrackStatistics to true after the cache is already created? The code appears to be checking the live value on every operation, but these fields may not have been initialized., right?

halter73 replied:
Should the other `if (_options.TrackStatistics)` be changed to `_allStats is not null` or something? I don't think that we should support changing `TrackStatistics` to true after the cache is already created. That's not really the intent of the option being settable. It's that way to support the "options pattern."

maryamariyan replied:
Switched the flag. TrackStatistics is not expected to change for the lifetime of the cache.

maryamariyan committed:
change flag condition to _allStats not null

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.csproj:
```diff
@@ -6,6 +6,10 @@
 
   <ItemGroup>
     <Compile Include="Microsoft.Extensions.Caching.Abstractions.cs" />
+    <Compile Include="Microsoft.Extensions.Caching.Abstractions.net7.0.cs" Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))" />
```
Why not also allow the DIM on net6?

maryamariyan replied:
It seems to be what we approved it with arbitrarily with `#if NET_7_OR_GREATER` over the API review meeting.

eerhardt replied:
I'm not sure how much it matters, and maybe we can wait for feedback, but I guess I would have added it for all the .NETCOREAPP TFMs we support. That way if someone was on the LTS .NET 6, but using the 7.0 MemoryCache nuget package, they could use this API.

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,98 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        private readonly List<WeakReference<Stats>?>? _allStats;
```
Can these be moved to be with the other fields in this class?

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,98 @@ void ScheduleTask(DateTimeOffset utcNow)
+                    }
+                }
+
+                _allStats.TrimExcess();
```
Why TrimExcess during Sum()? Wouldn't it be more appropriate to do it during RemoveFromStats?

My reasoning is if no one calls GetCurrentStatistics(), the excess won't be trimmed. And also, you won't need to trim unless you removed something from the list - and typically, I can imagine removing will happen less often than calling GetCurrentStatistics().

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,98 @@ void ScheduleTask(DateTimeOffset utcNow)
+            }
+            ~Stats()
+            {
+                _memoryCache.RemoveFromStats(this);
```
If we never added this instance to the _memoryCache, should we also not remove from it?

Similarly, I wonder if the reference to the `MemoryCache` shouldn't be a WeakReference. Is there a case where one of these ThreadLocals could live longer than the MemoryCache, and it would be a memory leak?

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +325,98 @@ void ScheduleTask(DateTimeOffset utcNow)
+            long hits = 0, misses = 0;
+            lock (_allStats!)
+            {
+                hits += _accumulatedStats!.Hits;
```
You could not set the hits and misses to 0 to start with, and just set them here.

```C#
            long hits, misses;
            lock (_allStats!)
            {
                hits = _accumulatedStats!.Hits;
                misses = _accumulatedStats!.Misses;

```

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -53,6 +53,13 @@ public MemoryCache(IOptions<MemoryCacheOptions> optionsAccessor!!, ILoggerFactor
 
+            if (_options.TrackStatistics)
+            {
+                _allStats = new();
```
I'm updating our coding style guidelines in https://github.com/dotnet/runtime/pull/67061 to require the Type being used when the type isn't obvious from the left-hand side. (basically the converse of our `var` guidelines.)

```suggestion
                _allStats = new List<WeakReference<Stats>>();
```

maryamariyan committed:
Applies latest PR feedback

eerhardt reviewed (approved):
Looks good. I just had a few minor comments.

maryamariyan committed:
Second round feedback

maryamariyan committed:
support DIM on net6 as well

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/ref/Microsoft.Extensions.Caching.Abstractions.Typeforwards.netcoreapp.cs:
```diff
@@ -0,0 +1,6 @@
+
+// The compiler emits a reference to the internal copy of this type in our non-NETCoreApp assembly
+// so we must include a forward to be compatible with libraries compiled against non-NETCoreApp Microsoft.Extensions.Caching.Abstractions
+[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Runtime.CompilerServices.IsExternalInit))]
```
We shipped a public System.Runtime.CompilerServices.IsExternalInit in Microsoft.Extensions.Caching.Abstractions?

eerhardt replied:
No. But once you use `init` in a `netstandard2.0` library, and someone else references your library, Roslyn binds to the `IsExternalInitAttribute` in your assembly.

See discussion starting at https://github.com/dotnet/runtime/issues/61737#issuecomment-974271053

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +328,94 @@ void ScheduleTask(DateTimeOffset utcNow)
+                {
+                    if (wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += Interlocked.Read(ref stats.Hits);
+                        misses += Interlocked.Read(ref stats.Misses);
```
Technically, these only need to be interlocked on 32-bit.  In practice, the overhead is small enough and this called infrequently enough that it probably doesn't matter.

pentp replied:
These only need to use `Volatile.Read` (which will be emitted as a regular read on 64-bit platforms).

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +328,94 @@ void ScheduleTask(DateTimeOffset utcNow)
+
+        internal sealed class Stats
+        {
+            public long Hits;
+            public long Misses;
+            private readonly MemoryCache? _memoryCache;
+            public Stats() { }
+            public Stats(MemoryCache memoryCache)
+            {
+                _memoryCache = memoryCache;
+                _memoryCache.AddToStats(this);
+            }
+            ~Stats()
+            {
+                _memoryCache?.RemoveFromStats(this);
+            }
```
Nit: some blank lines separating things would be nice:
```suggestion
            private readonly MemoryCache? _memoryCache;
            public long Hits;
            public long Misses;

            public Stats() { }

            public Stats(MemoryCache memoryCache)
            {
                _memoryCache = memoryCache;
                _memoryCache.AddToStats(this);
            }

            ~Stats() => _memoryCache?.RemoveFromStats(this);
```

maryamariyan committed:
Update src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

adamsitnik commented on src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheStatistics.cs:
```diff
@@ -0,0 +1,39 @@
+        public long? CurrentEstimatedSize { get; init; }
+
+        /// <summary>
+        /// Gets the total number of cache hits.
```
```suggestion
        /// Gets the total number of cache misses.
```

adamsitnik commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +328,94 @@ void ScheduleTask(DateTimeOffset utcNow)
+            }
+        }
+
+        private void Hit()
```
This method is on the hot path, it would be nice if it was getting inlined when stats are disabled. @maryamariyan can you see it in the profiles when stats are disabled?

If it does not get inlined we could consider forcing it to inline by using the attribute.

adamsitnik reviewed (approved):
LGTM, I believe that we should merge it now as it is (with stats disabled by default) and reconsider enabling after we receive feedback from customers.

maryamariyan committed:
Caching inlining

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:
```diff
@@ -295,6 +328,96 @@ void ScheduleTask(DateTimeOffset utcNow)
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Why do we have `NoInlining` here? I think that is the opposite of what @adamsitnik was asking for.

maryamariyan committed:
Removing method and writing code inline

maryamariyan commented:
Communicated offline with @noahfalk @adamsitnik and @sebastienros and we seem OK with merging this PR given that there would be an extension API to DI inject a memory cache with statistics tracking enabled by default (tracked in https://github.com/dotnet/runtime/issues/67769). But by default we will have statistics tracking off and as-is in the PR.

Also next for caching is to focus on having a good user experience for tracking multiple memory cache instances (tracked in https://github.com/dotnet/runtime/issues/67770), adding built-in metrics in the library and have a name to help differentiate between different statistics enabled caches. 

Merging.
